### ISSUE TITLE
Rainbow colors show up for Markdown files in visual mode but not source mode

### ISSUE THREAD

**Issue by ronblum** [rstudio/rstudio#7793]
### System details

    RStudio Edition : Desktop & Server
    RStudio Version : 1.4.804-1
    OS Version      : MacOS 10.15.6 & Red Hat 8.2
    R Version       : 4.0.2

### Steps to reproduce the problem

Create a Markdown file (not R Markdown) with the following chunk:
````
```{r rainbow}
([{([{}])}])
```
````

### Describe the problem in detail

In source mode for a Markdown file (not R Markdown), rainbow parens aren’t used. However, when switching to visual mode, rainbow parens are used.

### Describe the behavior you expected

Preferably, rainbow colors will be used, as they do in visual mode and in R Markdown files.

- [x] I have read the guide to submitting good bug reports at https://github.com/rstudio/rstudio/wiki/Writing-Good-Bug-Reports . 
- [x] I have installed the latest version of RStudio and confirmed that the issue still persists.
- [ ] If I am reporting a RStudio crash, I have included a diagnostics report. https://support.rstudio.com/hc/en-us/articles/200321257-Running-a-Diagnostics-Report
- [x] I have done my best to include a minimal, self-contained set of instructions for consistently reproducing the issue.

**Comment by mikebessuille** [comment]
rainbow parens is not meant for publishing, so we do not expect this to work in a Markdown file where the code is not being edited for execution.

**Comment by ronblum** [comment]
@mikebessuille Rainbow parens show up in visual mode for Markdown files, so the source and visual mode are inconsistent. Should they be disabled in visual mode?

**Comment by mikebessuille** [comment]
in that case, perhaps they should be disabled in Visual Mode, not sure.  Yes I think they should be consistent between visual mode and "regular mode".

**Comment by jjallaire** [comment]
Code in markdown files is indeed executed so needs to have identical behavior as source mode.

I can't quite follow what the defect is here (i.e. what is inconsistent). If rainbow parens are enabled they should show up in both source and visual mode (and conversely not show up when disabled).

**Comment by ronblum** [comment]
The difference is that all parens remain a single color in source mode, but in rainbow color in visual mode. E.g.

````
```{r}
(((((((())))))))
```
````
appears is monochrome in source mode:
<img width="190" alt="monochrome" src="https://user-images.githubusercontent.com/31009952/93813293-de7b5880-fc20-11ea-90a2-c6bd7b9aacf7.png">

but in rainbow colors in visual mode:
<img width="167" alt="rainbow" src="https://user-images.githubusercontent.com/31009952/93813460-24d0b780-fc21-11ea-891d-71a9a3d00f6b.png">

This is only true for Markdown. R Markdown is consistently rainbows.

**Comment by jjallaire** [comment]
@jmcphers Can we see what's going on here? I still don't understand from the bug description what the _setting_ actually is when you observe this behavior. This seems like it would be a P0 bug for preview (so I'll target it there for now) if rainbow parens were always displayed in visual mode no matter what the setting).

**Comment by jmcphers** [comment]
This is all presumably with rainbow parens turned on. @ronblum can confirm, but I think this is what we're looking at:

| Rainbow Setting | File Type | Mode | Result |
| -- | -- | -- | -- |
| Off | Markdown | Source | Off (OK) |
| Off | Markdown | Visual | Off (OK) |
| Off | R Markdown | Source | Off (OK) |
| Off | R Markdown | Visual | Off (OK) |
| On | Markdown | Source | Off **(BUG)** |
| On | Markdown | Visual | On (OK) |
| On | R Markdown | Source | On (OK) |
| On | R Markdown | Visual | On (OK) |

**Comment by ronblum** [comment]
That's correct—the problem, as described in the chart, only occurs in source mode when when rainbow parentheses are turned on: the parentheses are in rainbow colors in Markdown in visual mode, but are monochromatic in source mode.
![image](https://user-images.githubusercontent.com/31009952/93893541-bdb41100-fcbb-11ea-8f0c-ab9c7f6122dd.png)

When unchecked, there are (correctly) no rainbows in either mode.

**Comment by kevinushey** [comment]
> This seems like it would be a P0 bug for preview (so I'll target it there for now) if rainbow parens were always displayed in visual mode no matter what the setting).

Given that this isn't the case (it's only an issue in source mode) I think this can be moved out of v1.4.

**Comment by adamconroy** [comment]
I just checked, we do not currently do any R code coloring in regular Markdown files. Do we want to add that? This bug has nothing to do with rainbow parens.

![](https://i.imgur.com/tjt4TfU.png)
![](https://i.imgur.com/P8oi5cy.png)

**Comment by mikebessuille** [comment]
If that's the case, then perhaps this becomes an enhancement request, not a bug, because it's not just related to rainbow parens.  What is the behaviour of 1.3?

**Comment by ronblum** [comment]
The theme is not reflected for code in regular markdown files in 1.3. However, in 1.4, although the behavior is the same in the source mode, we do see the theme in visual mode. So, we have the same discrepancy between the two modes, as we is the case above for R markdown:

In a regular markdown file (not R markdown), enter the following code:
```{r cars}
summary(cars)
x <- 1
y <- 2
function(z)
install.packages("tidyverse")
```

Using the Dracula theme, the code is monochromatic in source mode:
<img width="241" alt="source" src="https://user-images.githubusercontent.com/31009952/94928959-6f102f00-0492-11eb-8580-593b675f8e61.png">

However, the code reflects the theme in visual mode:
<img width="274" alt="visual" src="https://user-images.githubusercontent.com/31009952/94929131-ada5e980-0492-11eb-8ffe-d174453e8ff5.png">

So what might have otherwise been an enhancement has arguably become a bug due to the difference between editor modes.

**Comment by mikebessuille** [comment]
Okay.  So this is a theme problem related to Visual Editor, and has nothing specifically to do with rainbow parens.  We should reflect that in the title, and we should fix this as part of the visual editor feature, in 1.4.

**Comment by ronblum** [comment]
Just to be clear, the colors work as expected in the visual editor. The problem is in the source editor.

**Comment by mikebessuille** [comment]
Right.  So, then it comes back to what I said earlier on this ticket.  This is not a regression; this never worked in source mode for Markdown files.  So... this wouldn't meet my bar at this point in the release for 1.4. In my opinion.  Anyone else feel strongly that it needs to be fixed urgently?

**Comment by kevinushey** [comment]
We've considered highlighting plain Markdown files as though they were R Markdown files in the past, but ultimately never decided to make that change.

FWIW, users can manually change the filetype using the file type menu at the bottom right of the editor viewport.

**Comment by ronblum** [comment]
The current state, in 1.4, has an inconsistency in markdown:
    - Highlighting and rainbow **are** present are present in visual editor
    - Highlighting and rainbow **are not** present and not in source mode.

Are we OK with this current state in 1.4? I say yes. If anyone disagrees, please say so.

**Comment by kevinushey** [comment]
Given that this only affects plain Markdown documents (and not R Markdown documents) I think it's okay to leave out of v1.4.
